deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -833,6 +850,50 @@ def increment_value(dictionary, key):
     def overlapping_partition_gradients_reduce_epilogue(self):
         self.independent_gradient_partition_epilogue()
 
+    def update_separate_grad_accum(self):
+        for group in self.bit16_groups:
+            for param in group:
+                if param.grad is not None:
+                    if param.grad_accum is None:
+                        param.grad_accum = param.grad.to(
+                            self.gradient_accumulation_dtype)
+                    else:
+                        param.grad_accum.add_(
+                            param.grad.to(self.gradient_accumulation_dtype).view(
+                                param.grad_accum.shape))
+                    # Don't clear if you still need the native grad for reduction (stage 2)
+                    if self.use_grad_accum_for_reduction:
+                        param.grad = None
+
+    def set_grad_accum_pointer(self):
+        for group in self.bit16_groups:
+            for param in group:
+                param.grad_accum = param.grad
+
+    # Clear the tensor the reduction gradient attribute is pointing to
+    def clear_grad_reduc_pointer(self, param):
+        param.grad_reduc = None
+        if self.use_grad_accum_for_reduction:
+            param.grad_accum = None
+        else:
+            param.grad = None
+
+    # Point to the gradient tensor to use in reduction
+    def set_grad_reduc_pointer(self, param):
+        if self.use_grad_accum_for_reduction:
+            param.grad_reduc = param.grad_accum
+        else:
+            param.grad_reduc = param.grad
+
+    # Called after a memory defrag moves tensor object via copy
+    def reset_pointers_to_reduc_after_copy(self, param):
```

Comment:
I think this could be replaced by 

```python
   def update_gradient_for_accumulation(self, param, new_grad):
        if ....: 
           param.grad_accum.copy_(new_grad)`
        else:
           param.grad = new_grad ``